gnocchi/amqp1d.py

```diff
@@ -0,0 +1,221 @@
+from gnocchi import utils
+
+LOG = daiquiri.getLogger(__name__)
+
+
+class BatchProcessor(object):
+    def __init__(self, conf):
+        self.conf = conf
+        self.incoming = incoming.get_driver(self.conf)
+        self.indexer = indexer.get_driver(self.conf)
+        self._ensure_resource_type_exists()
+
+        self._hosts = {}
+        self._measures = collections.defaultdict(
+            lambda: collections.defaultdict(list))
+
+    def reset(self):
+        self._hosts.clear()
+        self._measures.clear()
+
+    def treat_metric(self, host, name, value):
+        host_id = "%s:%s" % (self.conf.amqp1d.resource_type,
+                             host.replace("/", "_"))
+        self._hosts[host_id] = host
+        self._measures[host_id][name].append(incoming.Measure(
+            utils.dt_in_unix_ns(utils.utcnow()), value))
+
+    def flush(self):
+        try:
+            self._flush()
+        except Exception:
+            LOG.error("Unepected error during flush()", exc_info=True)
+        self.reset()
+
+    def _flush(self):
+        archive_policies = {}
+        for host_id, measures_by_names in six.iteritems(self._measures):
+            resource = self._get_resource(host_id)
+
+            names = list(measures_by_names.keys())
+            for name in names:
+                if name not in archive_policies:
+                    archive_policies[name] = (
+                        self.indexer.get_archive_policy_for_metric(name))
+            known_metrics = self.indexer.list_metrics(attribute_filter={
+                "and": [{"=": {"resource_id": resource.id}},
+                        {"in": {"name": names}}]
+            })
+            known_names = [m.name for m in known_metrics]
+            already_exists_names = []
+            for name in names:
+                if name not in known_names:
+                    try:
+                        m = self.indexer.create_metric(
+                            uuid.uuid4(),
+                            creator=self.conf.amqp1d.creator,
+                            resource_id=resource.id,
+                            name=name,
+                            archive_policy_name=archive_policies[name].name)
+                    except indexer.NamedMetricAlreadyExists as e:
+                        already_exists_names.append(e.metric)
+                    except indexer.IndexerException as e:
+                        LOG.error("Unexpected error, dropping metric %s",
+                                  name, exc_info=True)
+                    else:
+                        known_metrics.append(m)
+
+            if already_exists_names:
+                # Add metrics created in the meantime
+                known_names.extend(already_exists_names)
+                known_metrics.extend(
+                    self.indexer.list_metrics(
+                        names=already_exists_names,
+                        resource_id=resource.id)
+                )
```

Comment:
Watch out that they are not removed in the meantime, that could be another race condition here.